### PR TITLE
SQL Server: Index local temp tables

### PR DESCRIPTION
**Pull request by detule**
Fixes https://github.com/r-dbi/odbc/issues/584

With this, `dbExistsTable( conn, "#tempname")` should be reliable.  In turn writing to local temp tables with `dbWriteTable` or `dbAppendTable` should also be improved.

I am tracking additional issues with  temp tables in SQL Server.  Will address separately.

As I was working on this I was reminded that we need to add a DB/workflow using FreeTDS to CI/CD ( mostly a note to myself ).

Thanks!

### CODE REVIEW

File: R/db.R [hunk 1]
```diff
@@ -203,19 +215,92 @@ setMethod("sqlCreateTable", "DB2/AIX64",
...
-# both quotation marks as well as square bracket are used interchangeably for
-# delimited identifiers.  See:
-# https://learn.microsoft.com/en-us/sql/relational-databases/databases/database-identifiers?view=sql-server-ver16
-# Therefore strip the brackets first, and then call the DBI method that strips
-# the quotation marks.
-# TODO: the generic implementation in DBI should take a quote char as
-# parameter.
+#' SQL Server specific implementation.
+#'
+#' For SQL Server, conn@quote will return the quotation mark, however
```
Comment (hadley) [inline R/db.R:255]:
```suggestion
#' For SQL Server, `conn@quote` will return the quotation mark, however
```

File: R/db.R [hunk 2]
```diff
@@ -203,19 +215,92 @@ setMethod("sqlCreateTable", "DB2/AIX64",
...
+  args <- list(...)
+  if ( "catalog_name" %in% names(args) ) {
+    catalog_name <- args[["catalog_name"]]
+    if ( !is.null(catalog_name) &&
+        catalog_name != "%" &&
+        length(catalog_name ) > 0 &&
+        catalog_name != "tempdb" ) {
+      return(FALSE)
+    }
+  }
```
Comment (hadley) [inline R/db.R]: FWIW in cases like this, I think it's better to create a helper function and rely on R to do the dot matching, i.e. something like:

```R
mssq_is_temp <- function(..., name, catalog_name = NULL) {
  if ( !is.null(catalog_name) &&
        catalog_name != "%" &&
        length(catalog_name ) > 0 &&
        catalog_name != "tempdb" ) {
      return(FALSE)
    }
  ...
}

File: R/db.R [hunk 3]
```diff
@@ -203,19 +215,92 @@ setMethod("sqlCreateTable", "DB2/AIX64",
...
   })
+
+#' SQL Server specific implementation.
+#'
+#' Local temp tables are stored as
+#' \code{ [tempdb].[dbo].[#name]________(padding using underscores)[numeric identifier] }
+#'
+#' True if:
+#' - If catalog_name is supplied it must equal "temdb" or "%" ( wildcard )
+#' - Name must start with "#" followd by a non-"#" character
```
Comment (hadley) [inline R/db.R:281]: I'd suggest just writing all these method docs in a bulleted list in the `SQLServer` docs. That makes it a bit easier to see how the final rendered doc will look.

And I think it's fine to leave the `@usage` as is, so the read can more easily see what methods are customised.

Comment (detule) [reply R/db.R:281]: Thanks Hadley.

I took a look at the markdown in `SQLServer.md` and fwict it already looks like a bulleted list:

```
  \details{
  True if:
  \itemize{
  \item If catalog_name is supplied it must equal "temdb" or "\%" ( wildcard )
  \item Name must start with "#" followd by a non-"#" character
  }
```
Let me know if you had something else in mind.

RE: usage, I spend some time struggling with this:

```
❯ checking for code/documentation mismatches ... WARNING
  Functions or methods with usage in documentation object 'Microsoft SQL Server-class' but not in code:
    ‘\S4method{isTempTable}{Microsoft SQL Server,character}’
```

My first thought was that it has something to do with the class name having white-space characters but now i am not so sure.

The class and method are tagged like so:

```
  \name{Microsoft SQL Server-class}                                                                                                                                                                                                         
  \alias{Microsoft SQL Server-class}
  \alias{isTempTable,Microsoft SQL Server,character-method}
```
while the usage section:
```
\usage{
  \S4method{isTempTable}{`Microsoft SQL Server`,character}(conn, name, catalog_name = NULL, schema_name = NULL, ...)                                                                     
}
```

Let me know if you have any ideas.

Thanks again!

Comment (hadley) [reply R/db.R:281]: If you manually change the alias to the following, does the problem go away?

```
  \alias{isTempTable,`Microsoft SQL Server`,character-method}
```

Comment (detule) [reply R/db.R:281]: Good suggestion; alas doesn't work.

Comment (hadley) [reply R/db.R:281]: Weird! Definitely not worth looking into further.

**Review by hadley — APPROVED** [review]

**Pull request merged**